src/game/engine/terrain.cpp

```diff
@@ -0,0 +1,365 @@
+        if (!Unlimbo(Cell_To_Coord(cellnum))) {
+            delete this;
+        }
+    }
+    m_AnimStage.Set_Delay(0);
+}
+
+/**
+ * @brief
+ *
+ * @address 0x0056A270
+ */
+TerrainClass::~TerrainClass()
+{
+    if (GameActive) {
+        Limbo();
+    }
+}
+
+/**
+ * @brief
+ *
+ * @address 0x0056A3BC
+ */
+void *TerrainClass::operator new(size_t size)
+{
+    TerrainClass *this_ptr = g_Terrains.Alloc();
+    DEBUG_ASSERT(this_ptr != nullptr);
+    if (this_ptr != nullptr) {
+        this_ptr->m_IsActive = true;
+    }
+    return this_ptr;
+}
+
+/**
+ * @brief
+ *
+ * @address 0x0056A3DC
+ */
+void TerrainClass::operator delete(void *ptr)
+{
+    TerrainClass *this_ptr = static_cast<TerrainClass *>(ptr);
+    DEBUG_ASSERT(this_ptr != nullptr);
+    if (this_ptr != nullptr) {
+        this_ptr->m_IsActive = false;
+    }
+    g_Terrains.Free(this_ptr);
+}
+
+/**
+ * @brief
+ *
+ * @address 0x0056AB98
+ */
+coord_t TerrainClass::Center_Coord() const
+{
+    DEBUG_ASSERT(m_IsActive);
+    return Coord_Add(m_Coord, m_Type->Get_UnkCoord());
+}
+
+/**
+ * @brief
+ *
+ * @address 0x0056AC98
+ */
+coord_t TerrainClass::Target_Coord() const
+{
+    return Coord_Add(Coord_From_Lepton_XY(0, -m_Height), Sort_Y());
+}
+
+/**
+ * @brief
+ *
+ * @address 0x0056A668
+ */
+MoveType TerrainClass::Can_Enter_Cell(cell_t cellnum, FacingType facing) const
+{
+    if (cellnum >= MAP_MAX_AREA) {
+        return MOVE_NO;
+    }
+    for (const int16_t *list = Occupy_List(); *list != LIST_END; ++list) {
+        if (m_Type->Is_Waterbound()) {
+            if (!Map[cellnum + *list].Is_Clear_To_Build(SPEED_FLOAT)) {
+                return MOVE_NO;
+            }
+        } else {
+            if (!Map[cellnum + *list].Is_Clear_To_Build(SPEED_TRACK)) {
+                return MOVE_NO;
+            }
+        }
+    }
+    return MOVE_OK;
+}
+
+/**
+ * @brief
+ *
+ * @address 0x0056A91C
+ */
+void TerrainClass::AI()
+{
+    ObjectClass::AI();
+    if (What_Type() == TERRAIN_MINE && !(g_GameFrame % (900 * Rule.Ore_Growth_Rate()))) {
+        Map[As_Cell(As_Target())].Spread_Ore(true);
+    }
+    if (m_AnimStage.Stage_Changed()) {
+        Mark(MARK_REMOVE);
+        if (m_Crumbling) {
+            if (m_AnimStage.Get_Stage() == Get_Build_Frame_Count(Get_Image_Data()) - 1) {
+                delete this;
+                Map.Zone_Reset((1 | 2 | 4)); // TODO, implement enum for zone flags.
+            }
+        }
+    }
+}
+
+/**
+ * @brief
+ *
+ * @address 0x004CDCA0
+ */
+coord_t TerrainClass::Sort_Y() const
+{
+    DEBUG_ASSERT(m_IsActive);
+    return Coord_Add(m_Coord, m_Type->Get_UnkCoord());
+}
+
+/**
+ * @brief
+ *
+ * @address 0x0056AB58
+ */
+BOOL TerrainClass::Limbo()
+{
+    if (!m_InLimbo) {
+        Map[Coord_To_Cell(m_Coord)].Clear_Occupant_Bit(OCCUPANT_TERRAIN);
+    }
+    return ObjectClass::Limbo();
+}
+
+/**
+ * @brief
+ *
+ * @address 0x0056AAB4
+ */
+BOOL TerrainClass::Unlimbo(coord_t coord, DirType dir)
+{
+    // is Theater_Allowed right?
```

Comment:
Looks okay, so you can remove this comment here as it shouldn't end up in the final commit.